### ISSUE TITLE
Precompiled maps for Germany are corrupt as of Dec 30

### ISSUE THREAD

**Issue by mvglasow** [mapsforge/mapsforge#917]
The precompiled maps for Germany (both the full map and parts of it) on the download server are corrupt. The entire map of Germany is a mere 21M in size, while the map I downloaded six months ago was around 1.6G.

Original bug report here: https://github.com/mvglasow/satstat/issues/112

Seems to be a case of https://github.com/mapsforge/mapsforge-mapcreator/issues/12 – can you trigger a rebuild of the German map? Thanks!

**Comment by devemux86** [comment]
It was mentioned in the [forum](https://groups.google.com/d/msg/mapsforge-dev/5IrSdKKWDFM/JOLJqbS4DwAJ) and so the https://github.com/mapsforge/mapsforge-mapcreator/issues/12 came up.

@ludwigb restarts the server process for all maps to rebuild again.

**Comment by applantation** [comment]
Sorry I forgot to fix this over new year.
I will look into it sometime, but it will be a while before correct maps
appear again.
I know this is annoying ... also for me.

**Comment by devemux86** [comment]
>BTW I suppose that in multilingual/ there are as well files which are too short (or are the multilingual files only addons for the main maps?)

Multilingual maps need a refresh too.
They are full maps (not addons) built with extra [preferred-languages](https://github.com/mapsforge/mapsforge/blob/master/docs/Getting-Started-Map-Writer.md#basic-options) in map writer.

**Comment by applantation** [comment]
OK, everything restarted. I hope this goes ok this time, but it will take
time. Sorry.

**Comment by ludwigb** [comment]
Rebuild has been started, but there is no good way to rebuild selectively.
(The problem seems to be that the intermediate extracts are truncated,
which take the most time to generate)

**Comment by hgoebl** [comment]
What server resources do someone need to make a build?
Memory, HD-space, CPU, network bandwidth?
Is the problem a matter of memory shortage?
How often do you recreate the map-files?
Maybe outsourcing the process to cloud services might be a solution.
I'd like to help you if I can.

**Comment by devemux86** [comment]
Have a look also in [MapCreator](https://github.com/mapsforge/mapsforge-mapcreator) main page.

**Comment by jw243** [comment]
The problem is still not fixed. In the download section of SatStat, there are still only the currupt maps of Germany from December 30th available.

**Comment by hgoebl** [comment]
@jw243 that's true, but compilation is making progress. America (North+South), Australia, Africa is already built, excluding africa.map. AFAIK Russia about half of the maps. It's taking quite long time and hopefully it will succeed. :+1: I'm also looking forward to having the maps of Europe, but it seams to be the last item of the batch.

**Comment by devemux86** [comment]
Building the world is a time consuming process. :slightly_smiling_face: 

Regarding server use inside apps please see also in [forum](https://groups.google.com/d/msg/mapsforge-dev/R0pd_ayX124/w_IAiniaAgAJ).

**Comment by hgoebl** [comment]
It is said that god built the world in about six or seven days ;-)
The sad thing is that the mirror successfully fetched all the broken maps.
Is there a way to validate the created maps before being published and copied?
Can I help (Java, JavaScript, little bash)?

**Comment by devemux86** [comment]
We need to let the current cycle complete so all maps come again in good condition.

In next cycle we could try some changes, e.g. @ludwigb mentioned to skip continents intermediate extraction.

For the time we have open the https://github.com/mapsforge/mapsforge-mapcreator/issues/11 about filling the missing regions.
Any help in that is most welcome! :slightly_smiling_face: 

Also https://github.com/mapsforge/mapsforge-mapcreator/issues/12 about avoid overwriting good maps with bad ones.

**Comment by hgoebl** [comment]
For me it seams like if the process has been restarted again beginning with Australia or the like.
Can anybody check how's the server doing?
I've studied the code, but unfortunately I'm quite far away from being useful (no Python knowledge, no big computer to test the scripts; I'm stucking at installing shapely not finding geos_c; don't know how to install GDAL, ... sorry).

Would it help when I just split the config.xml into smaller parts?

And maybe one could skip the really fat map files (africa.map, australia-oceania.map) if they just exist for convenience and are already covered by smaller maps?

Another idea: what about a command line option to skip map-files which are younger than X (e.g. 1.1.2017)? This would be a way to easier resume from failed execution.

Or maybe "dryrun" could create a shell-script which can be executed in bash and cut down to particular parts easily (e.g. by `grep`)?

Sorry for being a bigmouth without actually helping you, but it begins to hurt somehow not having the valuable maps accessible.

**Comment by devemux86** [comment]
We need all good thoughts!

BTW the intermediate .pbf are not only for building big countries, but also for speed up the creation of their child areas, controlled via `create-pbf="true"` in [config.xml](https://github.com/mapsforge/mapsforge-mapcreator/blob/master/xml/example-config.xml).

See also my idea for a different build process in https://github.com/mapsforge/mapsforge-mapcreator/issues/13.
And @wish7code's proposal in https://github.com/mapsforge/mapsforge-mapcreator/issues/12#issuecomment-274276550.

**Comment by ludwigb** [comment]
Yes, the process restarted again. I used to run it twice per month, but
maybe now the whole process takes much longer.
But the Germany file failed because of some other reasons (as posted,
before the process was restarted).

I will cancel the next run, so we will have a bit of time to reevaluate the
situation after this run is completed.

I have not written these scripts either, and yes, the problem is that they
are difficult to test, as everything takes so long. There is also too much
logging output that is less than useless as it obscures errors, also
logging broken into too many files....

On the server I already broke the config files into smaller parts, partly
to execute in parallel, maybe even that is a mistake as it of course
doubles the memory usage. (It was fine for a while, but now as other people
have noticed the map files for big countries have grown so much, mostly
with building imports I guess).

Ludwig

**Comment by applantation** [comment]
I am currently trying to fix the script on the server.

I am taking the following actions:

   - I will internally revert back to a single process at a time
   (previously I was running two map building processes at the same time,
   maybe this was now causing build problems with memory)
   - I put a check so that only larger files are copied (to avoid the
   corruption problem, my approach is to copy as long as the new file is at
   least 80% of the size of the old one, this will cause problems if we e.g.
   decide to drop buildings from files to reduce the file size of maps like
   germany)
   - I put a check in so that a file that was built less than a month ago
   will not be rebuilt (in order to speed re-runs of the process up)

For the moment being I will be using a new directory
http://download.mapsforge.org/maps/2017/ on the server. I rsynced the old
files into that directory to test out if the additional checks work as
intended. As everyone who has looked at the files knows the scripts are
tricky, so at least new files will go to a new area)

Ludwig

**Comment by devemux86** [comment]
> I will internally revert back to a single process at a time (previously I was running two map building processes at the same time, maybe this was now causing build problems with memory)

:+1:

**Comment by devemux86** [comment]
@ludwigb there is a double 2017 in poi folders: http://download.mapsforge.org/pois/2017/2017/

**Comment by ludwigb** [comment]
Yes, I was trying to change the download process as that might involve new
error situations.

My check for file size is just a little more complicated, I will also
overwrite if the new file is a bit smaller (just to allow for changes in
data, e.g. an import being removed by OSM)

Ludwig

**Comment by devemux86** [comment]
Will you update [MapCreator](https://github.com/mapsforge/mapsforge-mapcreator) scripts after a successful run?
(so we can have a common base for future improvements)
